ConditionalCompilation.md

```diff
@@ -1,28 +1,140 @@
+
+Two new attributes are introduced that take a single parameter.
+* The `@ifdef` attribute marks the decorated node for removal if the parameter evaluates to `false`.
+* The `@ifndef` attribute marks the decorated node for removal if the parameter evaluates to `true`.
+
+*Example*
+
+```rs
+@ifdef(my_feature)
+fn f() { ... }
+@ifndef(my_feature)
+fn f() { ... }
+```
+
+### Syntax 2: `@if` attribute with compile-time logical expressions
+
+> This is the medium-complexity implementation.
+
+A *compile-time expression* is a subset of normal WGSL [expressions](https://www.w3.org/TR/WGSL/#expressions). it must be one of:
+ * a *compile-time feature*,
+ * a [logical expression](https://www.w3.org/TR/WGSL/#logical-expr): logical not (`!`), short-circuiting AND (`&&`), short-circuiting OR (`||`),
+ * a [parenthesized expression](https://www.w3.org/TR/WGSL/#parenthesized-expressions).
+
+Three new attributes are introduced.
+* An `@if` attribute takes a single parameter. It marks the decorated node for removal if the parameter evaluates to `false`.
+* An `@elif` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It takes one parameter.
+   It marks the decorated node for removal if its parameter evaluates to `true` AND if the previous `@if` and `@elif` attribute parameters evaluate to false.
+* An `@else` attribute decorates the next sibling of a syntax node decorated by a `@if` or an `@elif`. It does not take any parameter.
+   It marks the decorated node for removal if the previous `@if` and `@elif` attribute parameters evaluate to false.
+
+*Example*
+
+```rs
+@if(feature_1 && (!feature_2 || feature_3))
+fn f() { ... }
+@elif(!feature_1)
+fn f() { ... }
+@else
+fn f() { ... }
+```
+
+### Execution of the conditional compilation phase
+
+The conditional compilation phase *should* be the first phase to run in the full WESL compilation pipeline.
+
+1. The compiler is invoked with a list of features to enable.
+2. The source file is parsed.
+3. *Compile-time attributes* are evaluated
+    * If the syntax node is marked for removal: it is eliminated from the source code along with the attribute.
+    * Otherwise, only the attribute is eliminated from the source code.
+4. The updated source is passed to the next compilation phase. (e.g. import resolution) 
+
+### Possible future extensions
+
+* High-complexity *compile-time expressions*: if we end up implementing other compile-time attributes, such as loops (e.g. `@for`, `@repeat`), or [compile-time-evaluable](https://zig.guide/language-basics/comptime/) expressions, then we would need to extend the grammar of *compile-time expressions*. It would also affect this proposal.
+
+*Example*
+
+```rs
+@comptime
+fn response_to_the_ultimate_question() -> u32 {
+  return 42;
+}
+
+@if(response_to_the_ultimate_question() == 42)
+fn f() { ... }
+```
+
+* Decorating other WESL language extensions: import statements could be decorated with *compile-time attributes* too.
+
```

Comment:
I was going to ask about that! I think some projects are doing this in the wild, but I don't know whether its a needed mechanism or if there are alternatives.

Reply:
It's a reasonably useful mechanism for say, having a specialized mobile GPU shader. (target platform)
Then one doesn't have to put a hundred `@if`s into their mobile GPU shader, but can instead say "only bother importing if we are on mobile".

Reply:
Not sure about this too and didn't give a lot of thought. So I'll keep it for future reflexion.

One constraint it would add is that the conditional compilation pass would **have to** run before import resolution.
But that does not sound like a big concern.

Reply:
Regarding the `@comptime`: Very useful, but sufficiently out of scope for the time being. I'd say that the alternative is letting those projects evaluate such conditions themselves in Javascript/Rust code. And that injecting code into shaders by using the linker's APIs (which are not part of the specification) is always an option.